## • 8. PCA -- Gotchas  {.unnumbered #pcaIII}


```{r}
#| echo: false
#| message: false
#| warning: false
library(tweetrmd)
library(knitr)
library(dplyr)
library(readr)
library(stringr)
library(DT)
library(webexercises)
library(kableExtra)
library(ggplot2)
library(tidyr)
source("../../_common.R") 
hybrid_zone_phenotypes_link <- "https://raw.githubusercontent.com/ybrandvain/datasets/refs/heads/master/clarkia_hz_phenotypes.csv"
sm_hz_phenotypes <- readr::read_csv(hybrid_zone_phenotypes_link) |>
    filter(site == "SM")|>
    select(id,subspecies, protandry = avg_protandry, herkogamy = avg_herkogamy, 
           petal_area = avg_petal_area,lat, lon, ele)
```

::: {.motivation style="background-color: #ffe6f7; padding: 10px; border: 1px solid #ddd; border-radius: 5px;"}

**Motivating Scenario:**
You’ve used PCA and know how it works, but now it's time to critically evaluate a PCA.  What should you watch out for?

**Learning Goals: By the end of this subchapter, you should be able to:**

1. **Make a checklist of things to consider when looking at a PCA** and know how to evaluate them critically. 

   * Know not to overinterpret PC axes with little explanatory power.  
   * Know to look out for a horseshoe shape suggesting non-linearity. 
   * Know to watch out for how missing data are handled.  
   * Know when data should (not) be scaled.  
   * Know why variables should not be redundant.  
   * Know that PCA goes after the variance in the data, so the data matter a lot.  <br>   

2. **Recognize that sampling efforts impact PCA results.**   

   * Know to consider evenness of sampling when interpreting results.  
   * Know that results of a PCA **DO NOT** generalize. 
:::

---

Now that we can run a PCA and know how they work, let's think hard about how to interpret PCA results. These expand on warnings I made in the [PCA quickstart](https://ybrandvain.quarto.pub/applied-biostatistics-summarizingdata/book_sections/ordination/pca.html) and/or in the [previous chapter](https://ybrandvain.quarto.pub/applied-biostatistics-summarizingdata/book_sections/ordination/pcaII.html#find-pc-values-by-adding-up-each-trait-weighted-by-its-loading), but now we can approach them with a bit of sophistication. I first start with a list of things to worry about. You should ask these questions of every PCA you see. Next I look more deeply into the idea that PCA goes after the variance in the data, so the data matter a lot.


## **P**roceed **C**autiously **A**head 

I previously listed some first things to do when you see a PCA. These include:      

- Understanding the structure of the data.   
- Connecting  this structure to a biological interpretation (but watch out for artifacts -- see below). 
- Thinking about what could have gone wrong in this PCA (I specifically noted to be wary of a horseshoe shape). 



### PCA evaluation checklist: 


<input type="checkbox"> **Could structure be an artifact?** </input>   
In large scale datasets (such as -omics, or cases in which many people collected data or data are stratified), some portion of the variance might be caused by artifacts like who collected the data, which run of an instrument was used, what time the data was generated etc. rather than the motivating idea.    

:::aside
Your life will be better if you try to randomly spread such potential artifacts randomly across your more exciting biological factors, rather than identifying their potential impact later.  
:::

*Before rushing to interpret results biologically always look into the boring explanations.* You can check for these issues by making exploratory data visualizations or evaluating associations between such boring variables (observer, batch, time of day etc) and PCs of interest.



---


<input type="checkbox"> **Was the interpretation of 2D shape justified?** </input>    
Sometimes PCA plots show compelling shapes — curves, clusters, gradients — that are tempting to interpret biologically. But be cautious: these shapes don’t always reflect complex biology. For example, [Luca Cavalli-Sforza](https://en.wikipedia.org/wiki/Luigi_Luca_Cavalli-Sforza) famously used PCA to study human genetic variation. While foundational, some of his interpretations of the shapes in PC space were unjustified — @novembre2008b found that these patterns can arise from much simpler processes. Similarly, 2D structure can arise from nonlinearity in just one dimension.

*When you see an interesting shape in a PCA plot, don’t jump straight to a complex biological story.* Ask if it could be generated by a simple process. 

--- 

<input type="checkbox"> **Could the way in which missing data were handled impact PCs?** </input>     
If you want to run a PCA with [missing data](https://ybrandvain.quarto.pub/applied-biostatistics-summarizingdata/book_sections/ordination/pcaII.html#find-pc-values-by-adding-up-each-trait-weighted-by-its-loading) you have some decisions to make. Ideally, no data are missing, but of course this is not always possible -- for example in genomic analyses, data are often missing because not all sites are sequenced in all individuals.   

- Assigning each missing data point to the mean for that trait is a common practice, but if some 
individuals are missing more data than others they will be brought towards the center of PC space. This can be misleading — it makes individuals with lots of missing data appear artificially close to the center of PC space, even if their true trait values are far from average. This  is a known for the types of population genomic analyses I do (@Yi2022).    
- Alternatively imputing the missing data can give us more confidence in the shape of our PCs than we deserve and could mislead us by assigning wrong values to an individual's traits.   

One simple check is to plot the value of a PC on the y-axis against the proportion of missing data on the x-axis. Any linear or nonlinear trend is worrying.

---

<input type="checkbox"> **Did you do something silly?** </input>  This is somewhat embarrassing but I almost added `id` as a variable in this PCA. Of course [`id` should not be numeric](https://ybrandvain.quarto.pub/applied-biostatistics-preface/book_sections/example_mini/varType.html#types-of-variables), and is meaningless. So if I made this mistake and got super lucky there is just some noise in PC1. But it could be way worse, if id non-randomly assigned, it gets a meaningful weighting in PC space and we get super confused. So be sure you don't make any such silly mistakes. 

---

<input type="checkbox"> **Should data (not) be scaled?** </input>      

We discussed this at length in the [previous section](https://ybrandvain.quarto.pub/applied-biostatistics-summarizingdata/book_sections/ordination/pcaII.html#why-we-usually-scale-our-variables) — we usually scale variables so that differences in scale and variability don’t mislead us. But do we always want to give equal weight to variables with very different amounts of variance?

Take a PCA based on SNPs: scaling the data means that variation at a locus with a very rare allele is treated as just as important as variation at a highly polymorphic locus. That might not be a good idea. In some cases — like when differences in variance reflect meaningful biological differences, or when variables are on the same scale already — you may want to skip scaling. But be cautious -- @lever2017 suggest that when variables are similar, we should not scale the data.

---

<input type="checkbox"> **Are variables redundant?** </input>      
PCA is most useful when variables are correlated, since it finds axes that summarize that structure.  that capture the structure of correlations in the data. **But data should not be redundant.** 

For example the full *Clarkia* dataset contains measurements of both petal area and petal circumference. Such variables are very similar, so putting them both into a PCA is essentially double-counting flower size.  This redundancy means that a PCA will allocate more variance to the shared dimension, exaggerating its importance.  Consider reducing highly correlated traits to one value (e.g. their mean) or removing the less interpretable variable. How correlated is too correlated? There is  no hard and fast answer, but I suggest considering collapsing or removing variables when the absolute value of their correlation, $|r|$, exceeds 0.9.   


---

<input type="checkbox"> **What data went into the PCA and how even was sampling?** </input> 
PCA finds combinations of traits that best capture the variance **in the data**. Thus, is quite sensitive to the sampling scheme. I expand on this below!

## PCA depends on what you put into it   

When John Novembre showed that a PCA of genetic variation of Europeans looked a lot like a map of Europe, he knew what he was doing. He made a few specific decisions in his sampling and analysis, including --  making sure that all sampled individuals  had all four great grandparents from the same location, and that sampling was even across Europe. Had John sampled mostly Spaniards with a few other individuals spread around from across Europe, PCs would not make a map of Europe - but rather a somewhat distorted map of Spain. This is all to say that the details of sampling impact results of a principal component analysis. 

Just as sampling matters, so does the choice of what variables go into the PCA. John used many genetic markers from across the genome. A PCA based on the ABO locus would find clusters of blood groups, while  a PCA based on morphology would look different in some other way. 

@fig-pc_sample uses data from a natural hybrid zone between *Clarkia xantiana* subspecies to illustrate the importance of sampling effort. Panels A-C show a PC-plot based off of three floral traits -- petal area, anther stigma distance, and the time between pollen release and ovule receptivity -- when different numbers of *parviflora* and *xantiana* are sampled.  The top left panel (@fig-pc_sample A) shows that with 45 samples of each subspecies from site SAW, PC1 captures 84.6% of the variance in the data. The three "*xantiana*" samples with  PC1 values near zero may be first generation hybrids.   @fig-pc_sample  shows that the variance attributable to PC1 decreases when sampling 45 *parviflora* and 3 *xantiana*, and @fig-pc_sample   C shows that this decreases further when sampling 3 *parviflora*  and 45 *xantiana*.  @fig-pc_sample D shows the average proportion variance explained by PC1 across different sampling efforts. This illustrates the broader point: PCA doesn’t tell us about populations, but about variation in the data we choose and how we sample.


```{r}
#| eval: FALSE
#| echo: false
#| # This is slow
library(patchwork)
library(tidyr)
annon <- function(X,return_PCA = FALSE){
  tmp_dat <- suppressMessages(read_csv("~/Downloads/clarkia_hz_phenotypes.csv"))|> 
    na.omit()|>
    filter(site == "SAW")|>
    group_by(id)|>
    slice_sample(n=1)|>
    ungroup()     |>
    filter(subspecies %in% c("P","X"))|>
    mutate(n = ifelse(subspecies == "P", X[1] ,X[2]))|>
    group_by(subspecies,site)|>
    group_modify(~ slice_sample(.,  n = .x$n[1]))|>
    ungroup()
  tmp_pca <- tmp_dat|>
    select(starts_with("avg"))|>
    mutate_all(scale)|>
    prcomp()
  if(!return_PCA){
    ans <- tmp_pca |>
    tidy(matrix = "pcs")|>
    slice(1)|>
    pull(cumulative)
      return(ans)}
   if(return_PCA){return(list(this_dat = tmp_dat, pca = tmp_pca))}
}


z<- crossing(a = c(1:4,seq(5,46,5)),
         b = c(1:4,seq(5,46,5)))
pc_eff <- apply(z, 1,function(X){
  print(X)
  pc1<- replicate(n=20, annon(X))
  tibble(n_parv = X[1], n_xan = X[2],mean_PC1 = mean(pc1,na.rm=TRUE))}
)
 


PC_heatmap <- bind_rows(pc_eff)|>
   filter(mean_PC1<1)|>
        ggplot(aes(x = factor(n_parv), y = factor(n_xan), fill = mean_PC1))+
        geom_tile()+
        scale_fill_viridis_c(breaks = seq(.5,.9,.2),labels = c(".5",".7",".9"))+
  labs(x = expression("#"~italic("parviflora")),
       y = expression("#"~italic("xantiana")),
       fill = "Variance\nexplained\nby PC1",
       title = "Sampling balance influences PCs")+
  annotate(geom= "text", label = c("B","C","A"), x = c(13,3,13),y=c(3,13,13), 
           color = "red",fontface = "bold")+
  theme(legend.position = "none",
        axis.text = element_text(size = 11),
        axis.title = element_text(size = 15)) + 
  theme(legend.text = element_text(size = 13),
    legend.title = element_text(size = 12),
    legend.position = "inside",
    legend.position.inside = c(.62,.75), legend.direction = "horizontal",
    legend.key.width = unit(.4,"cm"))


more_xan <- annon(X = c(3,45),return_PCA = TRUE)
plot_1 <- autoplot(more_xan$pca,
                   data = more_xan$this_dat, color = "subspecies",
         size = 7, shape = 'subspecies')+
  scale_shape_manual(values = c("P","X"))+
  labs(title = expression(paste(3~italic(parviflora),","~45~italic(xantiana))))+
  theme(legend.position = "none",
        axis.text = element_text(size = 11),
        axis.title = element_text(size = 15))

more_parv <- annon(X = c(45,3),return_PCA = TRUE)
plot_2 <- autoplot(more_parv$pca,
                   data = more_parv$this_dat, color = "subspecies",
                   size = 7, shape = 'subspecies')+
  scale_shape_manual(values = c("P","X"))+
  labs(title = expression(paste(45~italic(parviflora),","~ 3~italic(xantiana))))+
  theme(legend.position = "none",
        axis.text = element_text(size = 11),
        axis.title = element_text(size = 15))
 
even_samp <- annon(X = c(45,45),return_PCA = TRUE)
plot_3    <- autoplot(even_samp$pca,
                      data = even_samp$this_dat, color = "subspecies",
         size = 7, shape = 'subspecies')+
  scale_shape_manual(values = c("P","X"))+
  labs(title = expression(paste(45~italic(parviflora),","~45~italic(xantiana))))+  
  theme(legend.position = "none",
        axis.text = element_text(size = 11),
        axis.title = element_text(size = 15))



(plot_3 + plot_2)/( plot_1 + PC_heatmap ) +
  plot_annotation('Relative sampling effort influences PCA',tag_levels = 'A',
                  theme = theme(plot.title = element_text(size = 15,face = "bold",
                                                          hjust = .5)))

```

```{r}
#| echo: false
#| label: fig-pc_sample
#| fig-cap: "**Sampling effort alters the results of PCA.** Panels A–C show principal component plots based on three floral traits in *Clarkia xantiana* (petal area, anther–stigma distance, and timing of pollen release vs. ovule receptivity). Data are from the “Sawmill Road” hybrid zone, with samples from both subspecies: *parviflora* is represented by red \"P\"s and *xantiana* by blue \"X\"s. Each panel presents a different sampling configuration: Panel A includes 45 individuals of each subspecies, Panel B includes 45 *parviflora* and 3 *xantiana*, and Panel C includes 3 *parviflora* and 45 *xantiana*. The proportion of total variance explained by PC1 decreases as the sampling becomes more uneven. Panel D summarizes this trend across combinations of sample sizes ranging from 1 to 45 for each subspecies. The color scale represents the average proportion of variance explained by PC1 across ten subsampling efforts for each parameter combination. Red letters A, B, and C correspond to Panels A–C."
#| fig-alt: "Four-panel figure showing how the results of PCA change with different sampling balances between two *Clarkia* subspecies. Panels A–C display PC1 vs. PC2 plots for three sampling configurations: equal sampling (45 each), and two imbalanced cases (45:3 and 3:45). As sampling becomes more imbalanced, the variance explained by PC1 decreases (84.62% → 73.01% → 55.99%). Panel D is a heatmap summarizing this effect across all sample combinations from 1 to 45 for each group, with brighter colors indicating higher variance explained by PC1. Locations of Panels A–C are marked within the heatmap."
include_graphics("../../figs/summarizing_data/ordination/pc_sample.png")
```

